### PR TITLE
Remove dependences from Chisel core on Chisel utils

### LINKED ISSUE #164: Separation of Utils and Frontend
With macro-based source locators happening ( #147 ), there now needs to be a clean separation of the Chisel frontend (core data types - Bool, UInt, etc) and things which invoke the frontend (like utils - Mux, Cat, PopCount, and friends) because macros require definition and invocation to be in different compilation units.

I think it makes sense for Chisel utils be strictly dependent on Chisel frontend since it provides an example of how libraries can be written and structured (rather than invoking Chisel frontend internal functions directly, bypassing the source locator macro layer). If necessary, utils can also have transform macros for source locators, and may demonstrate multiple levels of source locators.

Anyways, the current problematic calls I've found are:
- `Data.toBits` is dependent on `Cat`
- `Vec.count` is dependent on `PopCount`
- `Vec.{indexWhere, lastIndexWhere}` are dependent on `PriorityMux`
- `Vec.onlyIndexWhere` is dependent on `Mux1H`

This issue was brought up about a year ago, during the Chisel3 code review, but that was before separate compilation needed to happen, so the consensus then was to leave it as-is. That's not necessarily a good option anymore. Thoughts for solving this?

### PR DESCRIPTION
**Pull request by aswaterman**
Partially resolves #164

### CODE REVIEW

File: src/main/scala/Chisel/SeqUtils.scala [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  }
+
+  /** Counts the number of true Bools in a Seq */
+  def count(in: Seq[Bool]): UInt = {
+    if (in.size == 0) {
+      UInt(0)
+    } else if (in.size == 1) {
+      in.head
+    } else {
+      count(in.slice(0, in.size/2)) + Cat(UInt(0), count(in.slice(in.size/2, in.size)))
```
Comment (sdtwigg) [inline src/main/scala/Chisel/SeqUtils.scala]: Avoid Cat?

File: src/main/scala/Chisel/SeqUtils.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      Mux(in.head._1, in.head._2, priorityMux(in.tail))
+    }
+  }
+
+  /** Returns data value corresponding to lone true predicate */
+  def oneHotMux[T <: Data](in: Iterable[(Bool, T)]): T = {
+    if (in.tail.isEmpty) {
+      in.head._2
+    } else {
+      val masked = in map {case (s, i) => Mux(s, i.toBits, Bits(0))}
```
Comment (sdtwigg) [inline src/main/scala/Chisel/SeqUtils.scala]: map followed by guaranteed unapply can be written as the (effectively equivalent) for-comprehension:
for((s, i) <- in) yield Mux(s, i.toBits, Bits(0))

File: src/main/scala/Chisel/SeqUtils.scala (hunk 1 above)
Comment (aswaterman) [inline src/main/scala/Chisel/SeqUtils.scala]: oops

File: src/main/scala/Chisel/SeqUtils.scala (hunk 2 above)
Comment (aswaterman) [inline src/main/scala/Chisel/SeqUtils.scala]: Yeah, sure, I just moved the code from ChiselUtil.  I wasn't trying to rewrite it.

**Comment by aswaterman** [comment]
PR updated to reflect @sdtwigg's line notes

**Pull request merged**